### PR TITLE
My Site Dashboard: List Scrolling positions are not maintained #16436

### LINKED ISSUE #16436: My Site Dashboard: List Scrolling positions are not maintained
### Actual behavior
My Site dashboard(Home and Menu tab) List view positions should be maintained after changing the bottom tabs

### Expected behavior
The Scrolling position of the user should be maintained even after switching to a new tab 

### Steps to reproduce the behavior
- Login to the App
- Open My Site tab and Scroll to a position on either tab 
- Navigate to Reader tab 
- Open My Site tab 
- Notice that the My Site(Home and Menu) tab scrolls to the top

##### Tested on [device], Android [version], WPAndroid [version]
Oneplus Nord, Android 11

### PR DESCRIPTION
**Pull request by gitstart**
Fixes #16436
My Site Dashboard: List Scrolling positions are not maintained after navigating to other pages

Solution : 
Removed the `adapter.registerAdapterDataObserver` method in `MySiteTabFragment.kt`

To test:
- Login to the App
- Open My Site tab and Scroll to a position on either tab
- Navigate to Reader tab
- Return to My Site tab

## Regression Notes
1. Potential unintended areas of impact
NONE

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested visually

3. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Fix [loom video](https://www.loom.com/share/54ebbbe7dd414d99918fd4b0ec2a0d3b)

Issue link https://github.com/wordpress-mobile/WordPress-Android/issues/16436

### CODE REVIEW

**Comment by AjeshRPai** [comment]
@gitstart Thanks for working on this. 

This maintains the scrolling positions across the tabs but when there is a new item in the list on the top, the recycler view doesn't scroll to the top to reveal the item. I believe that behaviour also should be satisfied. You can test that scenario with the following steps 

**Recyclerview doesn't scroll to the top item** 
- Make sure the Initial screen is the Menu tab( Go to me - > App Settings -> Initial screen -> Menu)
- Switch to a site without quick start card(Next steps card) - If its visible completing the quick start would do
- Go to the menu tab 
- Scroll to a specific position 
- Switch to a new site with Quick start in progress (Alternatively creating a new site will also do)
- Notice that the Menu tab maintains the scrolling positions and hence the new quick start is not visible if not scrolled to the top 

https://user-images.githubusercontent.com/17463767/188812512-effc99b3-b253-4c97-b5e5-64fe9b8fc84a.mp4

To summarize 
- If there are no new items on the top of the list then the scrolling positions should be maintained 
- If there are new items at the top then the scrolling positions need not be maintained and should be scrolled to the 0th position 

I hope this information is clear.  Let me know if you need any additional info.

**Review by AjeshRPai — CHANGES_REQUESTED** [review]
@gitstart Thanks for working on this. 

This maintains the scrolling positions across the tabs but when there is a new item in the list on the top, the recycler view doesn't scroll to the top to reveal the item. I believe that behaviour also should be satisfied. You can test that scenario with the following steps 

**Recyclerview doesn't scroll to the top item** 
- Make sure the Initial screen is the Menu tab( Go to me - > App Settings -> Initial screen -> Menu)
- Switch to a site without quick start card(Next steps card) - If its visible completing the quick start would do
- Go to the menu tab 
- Scroll to a specific position 
- Switch to a new site with Quick start in progress (Alternatively creating a new site will also do)
- Notice that the Menu tab maintains the scrolling positions and hence the new quick start is not visible if not scrolled to the top 

https://user-images.githubusercontent.com/17463767/188812512-effc99b3-b253-4c97-b5e5-64fe9b8fc84a.mp4

To summarize 
- If there are no new items on the top of the list then the scrolling positions should be maintained 
- If there are new items at the top then the scrolling positions need not be maintained and should be scrolled to the 0th position 

I hope this information is clear.  Let me know if you need any additional info.

**Comment by gitstart** [comment]
> @gitstart Thanks for working on this.
> 
> This maintains the scrolling positions across the tabs but when there is a new item in the list on the top, the recycler view doesn't scroll to the top to reveal the item. I believe that behaviour also should be satisfied. You can test that scenario with the following steps
> 
> **Recyclerview doesn't scroll to the top item**
> 
> * Make sure the Initial screen is the Menu tab( Go to me - > App Settings -> Initial screen -> Menu)
> * Switch to a site without quick start card(Next steps card) - If its visible completing the quick start would do
> * Go to the menu tab
> * Scroll to a specific position
> * Switch to a new site with Quick start in progress (Alternatively creating a new site will also do)
> * Notice that the Menu tab maintains the scrolling positions and hence the new quick start is not visible if not scrolled to the top
> 
>  Screen_Recording_20220907-123613_WordPress.Pre-Alpha.mp4 
> To summarize
> 
> * If there are no new items on the top of the list then the scrolling positions should be maintained
> * If there are new items at the top then the scrolling positions need not be maintained and should be scrolled to the 0th position
> 
> I hope this information is clear. Let me know if you need any additional info.

Thank you @AjeshRPai for reviewing the PR, and explaining the requirement in depth. 
I got your point, will do the requested changes and push the PR asap.

**Comment by gitstart** [comment]
Hello @AjeshRPai,
Requested changes pushed. 
Please take a look now, hope it fulfilled the requirement. 
Thank you! 🙌

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt [hunk 1]
```diff
@@ -176,7 +176,9 @@ class MySiteTabFragment : Fragment(R.layout.my_site_tab_fragment),
         adapter.registerAdapterDataObserver(object : AdapterDataObserver() {
             override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                 super.onItemRangeInserted(positionStart, itemCount)
-                recyclerView.smoothScrollToPosition(0)
+                if (itemCount == ONE_ITEM && positionStart == FIRST_ITEM) {
```
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt:179]: @gitstart I wonder whether the check `itemCount == ONE_ITEM` is needed or not. There might be some cases where we might have two items added to the top, the Quick start card(Next Steps Card) and the stats button when switching to a new site. So in that case, this functionality might not work as expected, in that case too we want the recycler view to be scrolled to the top. So I think the `positionStart == FIRST_ITEM` should be enough. Wdyt?. 🤔

**Review by AjeshRPai — APPROVED** [review]
@gitstart I have left a suggestion for you to address. Once that's taken care of, I will review the changes. 👍🏼

**Review by AjeshRPai — CHANGES_REQUESTED** [review]
Accidentally approved the changes with the previous review. Sorry. Requesting changes.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt (hunk 1 above)
Comment (gitstart) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt:179]: Thank you for reviewing @AjeshRPai,
Will implement the solution you suggest and get back on this asap. 😊

Comment (gitstart) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt:179]: Hello @AjeshRPai,
 
If we only leave `positionStart == FIRST_ITEM` the recycler view will always scroll to the top because when we navigate back to My Site (for example from the notification) `+16` elements are added to the recycler view from position `0`. The `itemCount == ONE_ITEM` check was put there to prevent this. 
We have tried to reproduce `There might be some cases where we might have two items added to the top, the Quick start card (Next Steps Card) and when switching to a new site but failed`. In our case when we change sites the stats button the recycler view needs to be showing the element at position 0 so that the toolbar is expanded. If you can provide steps to reproduce it, it will be very helpful. 

Thank you!

Comment (AjeshRPai) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt:179]: @gitstart I am sorry about this, I tried to reproduce the scenario and I found that this scenario doesn't exist. When we switch from a site having paid features(business plan) to a free plan or when you switch from a site in which you are author to a site in which you are admin, **activity** log and **backup** are the options which gets added and removed from the list. The **stats** list item is always shown. So the scenario I mentioned above is invalid. Sorry for the confusion.

>If we only leave positionStart == FIRST_ITEM the recycler view will always scroll to the top because when we navigate back to My Site (for example from the notification) +16 elements are added to the recycler view from position 0. The itemCount == ONE_ITEM check was put there to prevent this.

Got it. Thanks for clarifying it. I get it now. I think we are good. I am going ahead and triggering CI. Once the CI is successful, I will merge the PR 👍🏼

**Review by AjeshRPai — APPROVED** [review]
Everything LGTM. Thanks for fixing this issue. 👍🏼 🙌🏼 . Triggering the CI on this build, once that's successful, I will merge the PR
